### PR TITLE
Implement `on_epoch_end` method on `BatchedDataset`

### PR DESCRIPTION
**Pull request by oliverholworthy**
Relates to #722 (issue)

### Goals :soccer:

Support Tensorflow/Keras 2.10 with our `BatchedDataset`

### Implementation Details :construction:

- Implementing the `on_epoch_end` method on the `BatchedDataset`. Calling `DataLoader.stop`.
  - I think this is what was happening implicitly previously (a result of calling the `len` method being called - which also calls `DataLoader.stop`).

### Testing Details :mag:

Add a test where the same loader created with `BatchedDataset` is passed to `.fit` more than once.

### CODE REVIEW

**Comment by oliverholworthy** [comment]
One other thing I've noticed is that `len` is now being called more frequently in 2.10 (more than once per epoch). So the resetting of the iterator will happen more frequently. 

It seems that the reason for calling `.stop` in `__len__` maybe was added as a work-around for the implementation of `__getitem__` which ignores the `idx` argument and makes the iterator progress, and when `shuffle=True` (the default). Resetting the iterator when `len` is called hides this issue.

https://github.com/NVIDIA-Merlin/models/blob/5b38605f9a68e7925cb7d283fff1f6427d2616a2/merlin/models/tf/dataset.py#L316-L322

**Comment by oliverholworthy** [comment]
Opened another PR as an alternative to this implementation which should improve performance by removing the calls to `.stop` #727

**Review by marcromeyn — APPROVED** [review]

**Comment by oliverholworthy** [comment]
rerun tests

**Comment by oliverholworthy** [comment]
For some reason `tests/unit/tf/models/test_base.py::test_train_metrics_steps` is failing in Jenkins. I've been unable to reproduce this locally so far.

**Comment by oliverholworthy** [comment]
Hmm, that test (`tests/unit/tf/models/test_base.py::test_train_metrics_steps`) is now passing. Seems like we may have a flaky test.
